lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -126,6 +126,104 @@ Value createLLVMIntegerConstant(OpBuilder &builder, Location loc, short width,
   LLVM::createIndexConstant(rewriter, loc, this->getTypeConverter(),           \
                             __VA_ARGS__)
 
+LLVM::LLVMFuncOp getVprintfDeclaration(OpBuilder &builder) {
+  auto moduleOp = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  StringRef funcName("vprintf");
+  Operation *funcOp = moduleOp.lookupSymbol(funcName);
+  if (funcOp)
+    return cast<LLVMFuncOp>(*funcOp);
+
+  auto *context = builder.getContext();
+  auto retType = builder.getI32Type();
+
+  SmallVector<Type> argsType{
+      LLVM::LLVMPointerType::get(IntegerType::get(context, 8)),
+      LLVM::LLVMPointerType::get(IntegerType::get(context, 8))};
+  auto funcType = LLVM::LLVMFunctionType::get(retType, argsType);
+
+  ConversionPatternRewriter::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(moduleOp.getBody());
+
+  return builder.create<LLVMFuncOp>(UnknownLoc::get(context), funcName,
+                                    funcType);
+}
+
+void llPrintf(StringRef msg, ValueRange args, OpBuilder &builder) {
+
+  static const char formatStringPrefix[] = "printfFormat_";
+  assert(!msg.empty() && "printf with empty string not support");
+
+  auto *context = builder.getContext();
+  auto moduleOp = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  auto funcOp = getVprintfDeclaration(builder);
+
+  Type int8Ty = builder.getIntegerType(8);
+  Type int8Ptr = LLVM::LLVMPointerType::get(int8Ty);
+  Type int32Ty = builder.getIntegerType(32);
+  Value one = builder.create<LLVM::ConstantOp>(
+      UnknownLoc::get(context), int32Ty, builder.getI32IntegerAttr(1));
+  Value zero = builder.create<LLVM::ConstantOp>(
+      UnknownLoc::get(context), int32Ty, builder.getI32IntegerAttr(0));
+
+  unsigned stringNumber = 0;
+  SmallString<16> stringConstName;
+  do {
+    stringConstName.clear();
+    (formatStringPrefix + Twine(stringNumber++)).toStringRef(stringConstName);
+  } while (moduleOp.lookupSymbol(stringConstName));
+
+  llvm::SmallString<20> formatString(msg);
+  formatString.push_back('\0');
+  size_t formatStringSize = formatString.size_in_bytes();
+  auto globalType = LLVM::LLVMArrayType::get(int8Ty, formatStringSize);
+
+  LLVM::GlobalOp global;
```

Comment:
Is a specific name necessary here?
I wonder if LLVM constop could work here.

Reply:
not work for constop.